nixos/xserver: expose services.xserver.displayManager.importedVariables

JohnRTitor opened the pull request:
## Description of changes

This option allows user to import specific environment variables into systemd user environment, this is useful situations where opening links into another app does not work. This is frequently the case with flatpak apps.

~After this PR, we can utilise this option for example in [sway module](https://github.com/NixOS/nixpkgs/blob/a71e967ef3694799d0c418c98332f7ff4cc5f6af/nixos/modules/programs/wayland/sway.nix#L125) and [hyprland module](https://github.com/NixOS/nixpkgs/blob/a71e967ef3694799d0c418c98332f7ff4cc5f6af/nixos/modules/programs/wayland/hyprland.nix#L66), which employ other mechanisms for importing crucial variables like PATH, into the systemd environment.~

This option only affects XSessions and NOT Wayland.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fufexan reviewed (commented):
From what I can see, `importedVariables` already does all the work, it's just not visible. Maybe changing its visibility and setting the default to the value already set in `config` is enough.

JohnRTitor commented:
Indeed, though I did not want to touch the defaults. Sure.

JohnRTitor committed:
nixos/xserver: expose xserver.displayManager.importedVariables option

This is useful for users trying to import variables like PATH or XDG_DATA_DIRS
into systemd user environment when starting a graphical session.

fufexan reviewed (commented):
We should also move https://github.com/NixOS/nixpkgs/blob/2a2796c2bc87c5689383217d0a5b1cc293342aae/nixos/modules/services/x11/display-managers/default.nix#L228-L236 to the `default` value in the option definition.

JohnRTitor commented:
Won't moving them to default make them easily overridable, which can cause trouble, for example if someone just specifies:

```nix
services.xserver.displayManager.importedVariables = [ "PATH" ];
```

Would the default values be applied?

JohnRTitor commented:
And I tried testing with this, it does not work. Are the values even being applied in a Wayland session?

```
  programs.hyprland.systemd.setPath = lib.mkForce false;
  services.xserver.updateDbusEnvironment = true;
  services.xserver.displayManager.importedVariables = [
    "PATH"
    "XDG_DATA_DIRS"
  ];
```

Tried on the Hyprland version on nixos-unstable.

jtojnar commented:
Yes, this is only invoked by Xsession wrapper, which is not used by Wayland session.

JohnRTitor commented:
Can we implement something like this for Wayland, by running the same commands when starting a Wayland session?

ElvishJerricco commented:
> I am not sure if the default values will be applied then.

You're right, they won't. If you put things in the option's `default`, then any definition will override and wipe out those things.

JohnRTitor commented:
I am tempted to merge this as it is, even though it does not impact Wayland session like I originally intended it would.

All I did is make it visible and added an example here.

fufexan commented:
Perhaps we could make `importedVariables` readonly, and add an `extraVariables` user-settable option.

ElvishJerricco commented:
@fufexan Why is that better than just doing it this way? That's just a more complicated way of doing the same thing that also can't be overridden if someone wants to.

JohnRTitor commented:
That was my original intention with this PR, yes, but then I realised variables defined here and in other modules will be merged into a list by the module system. So is there even a need to? 
https://github.com/NixOS/nixpkgs/blob/560f0817fc6018a895d73d28a401d3a5cd132d38/nixos/modules/services/x11/display-managers/default.nix#L232-L240

fufexan commented:
I assumed the option was not visible for a reason (users should probably not change it). I have no issue with merging it as is.

jtojnar commented:
I am not convinced the [PR that introduced this option](https://github.com/NixOS/nixpkgs/pull/93764) was desirable in the first place. If a service needs certain environment variable, it should be passed to it explicitly.

Most services probably do not need `XDG_SESSION_ID`. I just checked Fedora (created user service that runs `env` and set `XDG_SESSION_ID` in `/etc/profile.d`) and it does not pass that to services either.

Also, we already have extremely bloated user service environment compared to Fedora – ¿perhaps from `pam_env` being used for `environment.sessionVariables`?

JohnRTitor commented:
> I am not convinced the [PR that introduced this option](https://github.com/NixOS/nixpkgs/pull/93764) was desirable in the first place. If a service needs certain environment variable, it should be passed to it explicitly.

I wanted to introduce this option as a workaround of https://github.com/NixOS/nixpkgs/issues/189851. 

Importing PATH to systemd, usually fixes it.

jtojnar commented:
I think `PATH` and `XDG_DATA_DIRS` might be one of those things that make sense to introduce unconditionally, to reproduce `/usr/bin` and `/usr/share` being available to user services on FHS distros out of the box.

But as we noticed above, this option will not help anyone – Wayland desktops will not run the Xsession wrapper, and the X-based desktops typically invoke [`dbus-update-activation-environment`](https://dbus.freedesktop.org/doc/dbus-update-activation-environment.1.html) themselves (which is what we probably would want instead of `systemd import-environment`).

A proper solution might be replacing the use of `pam_env` in `environment.sessionVariables` with systemd [environment generators](https://www.freedesktop.org/software/systemd/man/latest/systemd.environment-generator.html). That way, services would have the environment variables without needing to import anything.

JohnRTitor commented:
> A proper solution might be replacing the use of `pam_env` in `environment.sessionVariables` with systemd [environment generators](https://www.freedesktop.org/software/systemd/man/latest/systemd.environment-generator.html). That way, services would have the environment variables without needing to import anything.

Thanks for this, I'll need to get an idea on how to do this though, would love pointers from the systemd team!
